### PR TITLE
Adds scripts, Dockerfiles, and GitHub Actions CI for checking formatting and autoformatting

### PR DESCRIPTION
**Pull request by ilumsden**
Since I know there's been some issues with formatting (mostly due to my VSCode being stupid), I put together this PR to provide some new tools to help us with code formatting. This PR is heavily based on LLNL/Caliper#605, which I put together for the same reason.

This PR adds several scripts to the repo (all can be found in `scripts/formatting/`):
* `check-formatting.sh`: uses `clang-format` to check if the files in `src/` and `include/` are formatted correctly
* `autoformat.sh`: uses `clang-format` to autoformat the files in `src/` and `include/`
* `Dockerfile.check`: a Dockerfile that runs `check-formatting.sh` with `clang-format` v17.0.6
* `Dockerfile.format`: a Dockerfile that runs `autoformat.sh` with `clang-format` v17.0.6
* `check-formatting-docker.sh`: a script that builds and runs `Dockerfile.check`
* `autoformat-docker.sh`: a script that builds and runs `Dockerfile.format`

Additionally, this PR adds a new GitHub Actions workflow to check if code is formatted correctly. This script is essentially just a GitHub CI version of `Dockerfile.check`.

Finally, this PR adds the results of autoformatting the repo.

### CODE REVIEW

File: include/dyad/stream/dyad_stream_core.hpp [hunk 1]
```diff
@@ -17,11 +17,10 @@
 #error "no config"
 #endif
 
-#include <iostream>
-#include <string>
 #include <climits>
-
 #include <dyad/stream/dyad_params.hpp>
+#include <iostream>
```
Comment (hariharan-devarajan) [inline include/dyad/stream/dyad_stream_core.hpp:22]: Fix external before internal.

File: scripts/formatting/autoformat-docker.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+cd $SCRIPT_DIR
+cd ..
+cd ..
+
+echo "Bulding Docker image"
+docker build -t caliper-autoformat -f ./scripts/formatting/Dockerfile.format .
+
+echo "Running Docker container to autoformat code with clang-format 17.0.6"
+docker run --rm -v $(pwd):/home/jovyan --name caliper-autoformat-container caliper-autoformat
```
Comment (hariharan-devarajan) [inline scripts/formatting/autoformat-docker.sh:22]: caliper

**Comment by ilumsden** [comment]
- [ ] Add newline to end of each file
- [ ] Enforce ordering of includes
- [ ] Tweak formatting to preserve include groups and sort within each group
- [ ] Update docs to state that includes should be placed in newline-separated groups, which external groups (e.g., for C++ Standard Library and external tools) coming before internal groups
- [ ] Merge Dockerfiles into hierarchical Dockerfile (note: figure out how to run a specific tag for this type of Dockerfile)
   * This file shows how to do it for compose: https://github.com/hariharan-devarajan/hpc-aio/blob/dev/compose-dev.yaml. Need to figure out how to do this via CLI

**Review by hariharan-devarajan — CHANGES_REQUESTED** [review]
Lets address the comments we made during meeting.

**Comment by ilumsden** [comment]
@hariharan-devarajan all the requested changes to this PR are complete.

File: src/dyad/common/dyad_rc.h
Comment (wangvsa) [inline src/dyad/common/dyad_rc.h:1]: I'm curious why the tabs before the comments in lines 29-31 were removed , whereas they were added for 57-62 lines? Is this done by autoformatting?

Comment (ilumsden) [reply src/dyad/common/dyad_rc.h:1]: @wangvsa it seems to be basing the indent level off of groupings of fields. This is usually how clang-format does things. If you want, I can play around with clang-format settings to see if I can change that behavior.

**Pull request merged**